doc/languages-frameworks/python: update (#316591)
NixOS/nixpkgs: doc/languages-frameworks/python.section.md

    libxslt
  ];

  # tests are meant to be ran "in-place" in the same directory as src

>   doCheck = false;

Reviewer: Commit message says "enable check phase" :thinking:

Author: I guess the test ran in my imaginary machine.🤷

#### Using pytestCheckHook {#using-pytestcheckhook}

`pytestCheckHook` is a convenient hook which will substitute

> a [`checkPhase`](#ssec-check-phase) which runs `pytest`. This is also beneficial

Reviewer: > which will set up (or configure) `checkPhase` to run `pytest`.

I think that's simpler.

It is recommended to test packages as part of the build process.
Source distributions (`sdist`) often include test files, but not always.

Previously, by default the command `python setup.py test` is run as part of the
[`checkPhase`](#ssec-check-phase), but now it is necessary to pass a custom [`checkPhase`](#ssec-check-phase).

> In almost all cases, `pytestCheckHook` is used.

Reviewer: Were still defaulting to that, aren"t we? Also you don't need to pass a custom `checkPhase`, but rather a hook to set up a certain test framework.

Author: It is still `doCheck = true` by default, but the hook (python setup.py test) is not provided unless `format = "setuptools"`. This means that unless we add `pytestCheckHook` to the input or provide checkPhase, test will not run in the current builder.

How about this?
> Previously, by default the command `python setup.py test` is run as part of the [`checkPhase`](#ssec-check-phase), but now it is necessary to pass a test framework hook like `pytestCheckHook`.
It may be necessary to provide a custom checkPhase for packages that do not use a common framework.

Reviewer: > By default the `checkPhase` executes the deprecated `python setup.py test` command.

This is the status quo, tests will still try to run `python setup.py test`, when no hook is given. We *could* however change that in a huge treewide change.

> The best practice today is to pass a test hook (e.g. `pytestCheckHook`, `unittestCheckHook`) into `nativeCheckInputs`, that will reconfigure the `checkPhase` to use that particular test framework. Occasionaly it may  be required to provide a custom `checkPhase`, when the package does not rely on a common framework.

Author: Do you mean that `python setup.py test` will be executed since `format = "setuptools"` is currently the default?
If so, I agree with you.

Reviewer: Oh right, that is depended on the format :facepalm:

* Python applications live outside of `python-packages.nix` and are packaged
  with [`buildPythonApplication`](#buildpythonapplication-function).
* Make sure libraries build for all Python interpreters.
  If it fails to build on some Python versions, consider disabling them by setting `disable = pythonAtLeast "3.x"` along with a comment.
* The two parameters, `pyproject` and `build-system` are set to avoid the legacy setuptools/distutils build.
* Packages with versioning like `pydantic_1` are not included in the `dependencies`.

>   Such packages should not be used in libraries, because propagation causes conflicts.

Reviewer: I think the takeaway could be something like this:

Only unversioned attributes (e.g. `pydantic`, but not `pypdantic_1`)  can be included in `dependencies`, since due to `PYTHONPATH` limitations we can only ever support a single version for libraries without running into duplicate module name conflicts.

> * The version restrictions of the dependencies are relaxed by [`pythonRelaxDepsHook`](#using-pythonrelaxdepshook).

Reviewer: ```suggestion
* The version restrictions of `dependencies` can be relaxed by [`pythonRelaxDepsHook`](#using-pythonrelaxdepshook).
```

> * Make sure the tests are enabled with [`pytestCheckHook`](#using-pytestcheckhook) or something and, in the case of

Reviewer: ```suggestion
* Make sure the tests are enabled using for example [`pytestCheckHook`](#using-pytestcheckhook) and, in the case of
```

  libraries, are passing for all interpreters. If certain tests fail they can be
  disabled individually. Try to avoid disabling the tests altogether. In any
  case, when you disable tests, leave a comment explaining why.

> * `pythonImportsCheck` is set. This is still a valid check even if `pytestCheckHook` is set.

Reviewer: ```suggestion
* `pythonImportsCheck` is set. This is still a good smoke test even if `pytestCheckHook` is set.
```

https://en.wikipedia.org/wiki/Smoke_testing_(software)
